### PR TITLE
Standardize failure case ordering

### LINKED ISSUE #500: solana_token_audit#10: Inconsistent error check order
Description
Method process_transfer checks for the MintMismatch case before the AccountFrozen one,
most other functions checks for AccountFrozen first.

Proof of Issue
Method process_transfer
Filename: token/program/src/processor.rs
Beginning Line Number: 171

Severity and Impact Summary
The inconsistency in checking for a frozen account could lead to undefined results if something
will interfere with the execution. Keeping a consistent structure on the way that the checks are
made make the implementation more predictable in its behavior.

Recommendation
Change the order so that AccountFreeze is checked before MintMismatch

### PR DESCRIPTION
**Pull request by CriesofCarrots**
While relevant failure cases vary from instruction to instruction, there are some that are consistent across many. It makes code easier to understand and test if these consistent failure cases are checked in the same order across instructions.

Here is the general order implemented:
1. Frozen
2. Account type is wrong in some way (native vs. non-native, authority type not applicable)
3. Insufficient funds
4. Mint mismatch
5. Decimal mismatch
6. Delegated checks

Fixes #500 

- [x] Rebase on #507

### CODE REVIEW

**Comment by jackcmay** [comment]
@CriesofCarrots Feel free to rebase now that the backports have gone in

File: token-swap/program/src/processor.rs [hunk 1]
```diff
@@ -2086,7 +2086,7 @@ mod tests {
             ) = accounts.setup_token_accounts(
                 &user_key,
                 &withdrawer_key,
-                initial_a,
+                withdraw_amount,
```
Comment (jackcmay) [inline token-swap/program/src/processor.rs:2089]: This change seems unrelated, in here by accident?

**Review by jackcmay — APPROVED** [review]

File: token-swap/program/src/processor.rs (hunk 1 above)
Comment (CriesofCarrots) [reply token-swap/program/src/processor.rs:2089]: No, CI fails without this change on `Err(Custom(1))` instead of the expected `Err(Custom(3))`. This is because I changed  the order of error-checks. Updating the initial_a amount ensures that we are making it to the mint check.
